### PR TITLE
improved binary encoding

### PR DESCRIPTION
**Pull request by missinglink**
Branched off [PR63](https://github.com/pelias/pbf2json/pull/63) see [diff](https://github.com/pelias/pbf2json/compare/centroid...binary_encoding?expand=1)

The code in this repo was written before I knew much about Golang, so some of the code is sub-optimal and needs a facelift.

This PR updates the formatting used to store nodes in leveldb, all nodes are stored so that they can be loaded for assembling the ways, some care must be taken to ensure we use the fewest bytes possible in order to keep the disk requirement low.

Prior to this PR the nodes were encoded as strings, the floats were truncated at 7 (6 before PR 63) decimal places and would consume:
- a best-case of 19 bytes for the coordinate (9 bytes per dimension + delimiter)
- a worst-case of 25 bytes for the coordinate (12 bytes per dimension + delimiter)
- a minimum of 19 bytes (two dimensions + delimiters + no metadata)
- a maximum of 29 bytes (two dimensions + delimiters + entrance & accessibility info)

After changing the encoding we get the full precision of 64 bits (very accurate) with the disk consumption of:
- a minimum of 16 bytes (8 bytes per dimension, no delimiter)
- a maximum of 17 bytes (including a bitmask)

The bitmask currently uses 2 bits for the entrance type and 2 bits for the wheelchair accessibility level (both uint8).

A further 4 bits are reserved for future use, it would also be simple to add additional bits to the bitmask in the future.

tl;dr we get a ~25% saving in disk space and also get increased float precision

### CODE REVIEW

**Comment by missinglink** [comment]
I was reading [this article](http://www.ctrl-alt-test.fr/2018/making-floating-point-numbers-smaller/) last night about the demoscene and how they manage to fix so much into only 64kb.

One of the techniques they use is float truncation where you zero or remove some of the rightmost bits of a float in order to reduce filesize (at the cost of precision).

So I did a little math and discovered we only need the first 6 bytes of a 64-bit float in order to achieve 8 digits of precision, far more than we need for geographic coordinates.

I added one more commit which reduces the storage per node down to 12 bytes, or 13 bytes including metadata.

**Comment by orangejulius** [comment]
Nice. now just add a way to generate some raytraced scenes and audio while the import is going

**Comment by orangejulius** [comment]
Looks good and I can confirm it drops disk usage of a Portland metro import from 125MB to 107MB, which is not bad.
